File: backend/app/schemas/human_review.py
"""
Pydantic Schemas for Part 8C: Human Review & Evidence Inspection Workflow
"""

import uuid
from datetime import datetime
from decimal import Decimal
from enum import Enum
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ReviewTypeEnum(str, Enum):
    COMPLIANCE_REVIEW = "COMPLIANCE_REVIEW"
    VERIFICATION_REVIEW = "VERIFICATION_REVIEW"
    DOCUMENT_REVIEW = "DOCUMENT_REVIEW"
    IDENTITY_MISMATCH = "IDENTITY_MISMATCH"
    LOW_CONFIDENCE = "LOW_CONFIDENCE"
    PENDING_SOURCE = "PENDING_SOURCE"
    CRITICAL_REVIEW = "CRITICAL_REVIEW"
    OTHER = "OTHER"


class ReviewSeverityEnum(str, Enum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class ReviewStatusEnum(str, Enum):
    OPEN = "OPEN"
    IN_REVIEW = "IN_REVIEW"
    RESOLVED = "RESOLVED"
    ESCALATED = "ESCALATED"
    SUPERSEDED = "SUPERSEDED"


class ReviewResolutionEnum(str, Enum):
    CONFIRMED = "CONFIRMED"
    REJECTED = "REJECTED"
    NEEDS_MORE_EVIDENCE = "NEEDS_MORE_EVIDENCE"
    ESCALATED = "ESCALATED"
    NOT_APPLICABLE = "NOT_APPLICABLE"


# -------------------------------------------------------------------------
# Queue Item & KPI Schemas
# -------------------------------------------------------------------------

class ReviewQueueItemResponse(BaseModel):
    id: uuid.UUID
    tender_id: uuid.UUID
    tender_number: str
    tender_title: str
    bid_id: uuid.UUID
    bid_number: str
    bidder_name: str
    bidder_pan: Optional[str] = None
    bidder_gstin: Optional[str] = None
    requirement_code: Optional[str] = None
    requirement_name: Optional[str] = None
    category: Optional[str] = None
    review_type: ReviewTypeEnum
    severity: ReviewSeverityEnum
    status: ReviewStatusEnum
    source_type: str
    title: str
    reason: str
    is_critical: bool = False
    is_mandatory: bool = False
    claimed_by_name: Optional[str] = None
    resolved_by_name: Optional[str] = None
    resolution: Optional[ReviewResolutionEnum] = None
    created_at: datetime
    resolved_at: Optional[datetime] = None


class ReviewQueueKPIs(BaseModel):
    total_open: int = 0
    critical_open: int = 0
    in_review: int = 0
    resolved_today: int = 0
    escalated: int = 0


class ReviewQueueResponse(BaseModel):
    kpis: ReviewQueueKPIs
    items: List[ReviewQueueItemResponse]
    total_count: int
    page: int
    page_size: int
    total_pages: int


# -------------------------------------------------------------------------
# Notes & Actions
# -------------------------------------------------------------------------

class ReviewNoteItem(BaseModel):
    id: uuid.UUID
    author_id: uuid.UUID
    author_name: str
    author_email: str
    author_role: str
    note_text: str
    created_at: datetime


class AddReviewNoteRequest(BaseModel):
    note_text: str = Field(..., min_length=2, max_length=5000, description="Reviewer remark or audit note")


class ResolveReviewRequest(BaseModel):
    resolution: ReviewResolutionEnum
    reason: str = Field(..., min_length=5, max_length=5000, description="Mandatory factual justification for review resolution")
    effective_compliance_status: Optional[str] = Field(
        None,
        description="Optional effective compliance status (PASS, FAIL, NOT_APPLICABLE) when resolving a compliance discrepancy",
    )


class StartReviewRequest(BaseModel):
    pass


# -------------------------------------------------------------------------
# Evidence Inspection Detail Sections
# -------------------------------------------------------------------------

class ReviewRequirementSection(BaseModel):
    requirement_id: Optional[uuid.UUID] = None
    code: str
    name: str
    category: Optional[str] = None
    requirement_type: Optional[str] = None
    expected_value: Any = None
    operator: Optional[str] = None
    is_mandatory: bool = True
    is_critical: bool = False
    weight: Optional[float] = 10.0


class ReviewActualEvidenceSection(BaseModel):
    claimed_value: Any = None
    verified_value: Any = None
    match_status: Optional[str] = None
    extraction_confidence: Optional[float] = None
    field_confidence: Optional[str] = None
    compliance_status: Optional[str] = None
    system_reason: Optional[str] = None


class ReviewSourceDocumentSection(BaseModel):
    document_id: Optional[uuid.UUID] = None
    document_name: Optional[str] = None
    document_type: Optional[str] = None
    file_size: Optional[int] = None
    content_type: Optional[str] = None
    uploaded_at: Optional[datetime] = None
    processing_status: Optional[str] = None
    page_number: Optional[int] = None
    extracted_text_snippet: Optional[str] = None
    ocr_confidence: Optional[float] = None
    secure_download_url: Optional[str] = None


class ReviewVerificationEvidenceSection(BaseModel):
    verification_record_id: Optional[uuid.UUID] = None
    verification_type: Optional[str] = None
    verification_status: Optional[str] = None
    registry_status: Optional[str] = None
    match_status: Optional[str] = None
    source_type: Optional[str] = None
    source_name: Optional[str] = None
    is_mock: bool = False
    is_available: bool = True
    confidence_score: Optional[float] = None
    evidence_payload: Optional[Dict[str, Any]] = None


class ReviewComplianceEvidenceSection(BaseModel):
    compliance_result_id: Optional[uuid.UUID] = None
    compliance_status: Optional[str] = None
    expected_value: Any = None
    actual_value: Any = None
    operator: Optional[str] = None
    reason: Optional[str] = None
    is_mandatory: bool = True
    is_critical: bool = False
    effective_compliance_status: Optional[str] = None
    human_resolution: Optional[str] = None
    human_reason: Optional[str] = None


class CrossDocumentComparisonRow(BaseModel):
    field_name: str
    pan_doc_value: Optional[str] = None
    gst_doc_value: Optional[str] = None
    mca_doc_value: Optional[str] = None
    other_doc_value: Optional[str] = None
    is_match: bool = True
    discrepancy_note: Optional[str] = None


class ReviewAICitationItem(BaseModel):
    citation_id: str
    source_type: str
    title: str
    page: Optional[int] = None
    snippet: Optional[str] = None


class ReviewAIExplanationSection(BaseModel):
    recommendation: Optional[str] = None
    confidence_label: Optional[str] = None
    summary: Optional[str] = None
    strengths: List[str] = Field(default_factory=list)
    concerns: List[str] = Field(default_factory=list)
    review_items: List[str] = Field(default_factory=list)
    grounded_citations: List[ReviewAICitationItem] = Field(default_factory=list)
    disclaimer: str = "AI explanation is advisory assistance and does not resolve this review item or make procurement decisions."
    is_stale: bool = False
    is_available: bool = True


# -------------------------------------------------------------------------
# Full Review Detail Workspace Response
# -------------------------------------------------------------------------

class ReviewDetailResponse(BaseModel):
    review_id: uuid.UUID
    organization_id: uuid.UUID
    tender_id: uuid.UUID
    tender_number: str
    tender_title: str
    bid_id: uuid.UUID
    bid_number: str
    bidder_legal_name: str
    trade_name: Optional[str] = None
    bidder_pan: Optional[str] = None
    bidder_gstin: Optional[str] = None
    review_type: ReviewTypeEnum
    severity: ReviewSeverityEnum
    status: ReviewStatusEnum
    title: str
    reason: str
    system_finding: Dict[str, Any] = Field(default_factory=dict)
    resolution: Optional[ReviewResolutionEnum] = None
    resolution_reason: Optional[str] = None
    effective_compliance_status: Optional[str] = None
    claimed_by_name: Optional[str] = None
    claimed_by_id: Optional[uuid.UUID] = None
    resolved_by_name: Optional[str] = None
    resolved_by_id: Optional[uuid.UUID] = None
    resolved_at: Optional[datetime] = None
    version: int = 1
    created_at: datetime
    updated_at: datetime

    requirement_section: Optional[ReviewRequirementSection] = None
    actual_evidence_section: Optional[ReviewActualEvidenceSection] = None
    source_document_section: Optional[ReviewSourceDocumentSection] = None
    verification_section: Optional[ReviewVerificationEvidenceSection] = None
    compliance_section: Optional[ReviewComplianceEvidenceSection] = None
    cross_document_section: List[CrossDocumentComparisonRow] = Field(default_factory=list)
    ai_explanation_section: Optional[ReviewAIExplanationSection] = None
    notes_history: List[ReviewNoteItem] = Field(default_factory=list)

File: frontend/src/types/human_review.ts
/**
 * Frontend TypeScript Types for Part 8C: Human Review & Evidence Inspection Workflow
 */

export type ReviewType =
  | "COMPLIANCE_REVIEW"
  | "VERIFICATION_REVIEW"
  | "DOCUMENT_REVIEW"
  | "IDENTITY_MISMATCH"
  | "ORGANIZATION_MISMATCH"
  | "LOW_CONFIDENCE"
  | "PENDING_SOURCE"
  | "CRITICAL_REVIEW"
  | "POTENTIAL_DOCUMENT_REUSE"
  | "DOCUMENT_REUSE_ALERT"
  | "CROSS_BIDDER_REUSE"
  | "POOR_DOCUMENT_QUALITY"
  | "EXPIRED_CERTIFICATE"
  | "BLACKLISTING_SIGNAL"
  | "UNRESOLVED_CLARIFICATION"
  | "OTHER";

export type ReviewSeverity = "LOW" | "NORMAL" | "MEDIUM" | "HIGH" | "CRITICAL";

export type ReviewStatus =
  | "OPEN"
  | "IN_REVIEW"
  | "IN_PROGRESS"
  | "AWAITING_CLARIFICATION"
  | "RESOLVED"
  | "DISMISSED"
  | "ESCALATED"
  | "SUPERSEDED";

export type ReviewResolution =
  | "CONFIRMED"
  | "CONFIRMED_BENIGN"
  | "CONFIRMED_REUSE"
  | "REJECTED"
  | "DISMISSED"
  | "NEEDS_MORE_EVIDENCE"
  | "ESCALATED"
  | "NOT_APPLICABLE";

export interface ReviewQueueItem {
  id: string;
  tender_id: string;
  tender_number: string;
  tender_title: string;
  bid_id: string;
  bid_number: string;
  bidder_name: string;
  bidder_pan?: string | null;
  bidder_gstin?: string | null;
  requirement_code?: string | null;
  requirement_name?: string | null;
  category?: string | null;
  review_type: ReviewType;
  issue_type_display?: string | null;
  severity: ReviewSeverity;
  status: ReviewStatus;
  source_type: string;
  title: string;
  reason: string;
  is_critical: boolean;
  is_mandatory: boolean;
  claimed_by_name?: string | null;
  resolved_by_name?: string | null;
  resolution?: ReviewResolution | null;
  created_at: string;
  resolved_at?: string | null;
}

export interface ReviewQueueKPIs {
  total_open: number;
  critical_open: number;
  high_open: number;
  awaiting_clarification: number;
  in_review: number;
  resolved_today: number;
  escalated: number;
}

export interface ReviewQueueResponse {
  kpis: ReviewQueueKPIs;
  items: ReviewQueueItem[];
  total_count: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ReviewNoteItem {
  id: string;
  author_id: string;
  author_name: string;
  author_email: string;
  author_role: string;
  note_text: string;
  created_at: string;
}

export interface AddReviewNoteRequest {
  note_text: string;
}

export interface ResolveReviewRequest {
  resolution: ReviewResolution;
  reason: string;
  effective_compliance_status?: string | null;
}

export interface ReviewRequirementSection {
  requirement_id?: string | null;
  code: string;
  name: string;
  category?: string | null;
  requirement_type?: string | null;
  expected_value?: any;
  operator?: string | null;
  is_mandatory: boolean;
  is_critical: boolean;
  weight?: number | null;
}

export interface ReviewActualEvidenceSection {
  claimed_value?: any;
  verified_value?: any;
  match_status?: string | null;
  extraction_confidence?: number | null;
  field_confidence?: string | null;
  compliance_status?: string | null;
  system_reason?: string | null;
}

export interface ReviewSourceDocumentSection {
  document_id?: string | null;
  document_name?: string | null;
  document_type?: string | null;
  file_size?: number | null;
  content_type?: string | null;
  uploaded_at?: string | null;
  processing_status?: string | null;
  page_number?: number | null;
  extracted_text_snippet?: string | null;
  ocr_confidence?: number | null;
  secure_download_url?: string | null;
}

export interface ReviewVerificationEvidenceSection {
  verification_record_id?: string | null;
  verification_type?: string | null;
  verification_status?: string | null;
  registry_status?: string | null;
  match_status?: string | null;
  source_type?: string | null;
  source_name?: string | null;
  source_badge_label?: string | null;
  is_mock: boolean;
  is_available: boolean;
  confidence_score?: number | null;
  evidence_payload?: Record<string, any> | null;
}

export interface ReviewComplianceEvidenceSection {
  compliance_result_id?: string | null;
  compliance_status?: string | null;
  expected_value?: any;
  actual_value?: any;
  operator?: string | null;
  reason?: string | null;
  is_mandatory: boolean;
  is_critical: boolean;
  effective_compliance_status?: string | null;
  human_resolution?: string | null;
  human_reason?: string | null;
}

export interface ReviewRiskSection {
  risk_level: string;
  risk_score?: number | null;
  top_signals: string[];
  is_critical: boolean;
}

export interface ReviewClarificationSection {
  clarification_id?: string | null;
  status?: string | null;
  status_label?: string | null;
  subject?: string | null;
  question?: string | null;
  response?: string | null;
  has_active_request: boolean;
}

export interface CrossDocumentComparisonRow {
  field_name: string;
  pan_doc_value?: string | null;
  gst_doc_value?: string | null;
  mca_doc_value?: string | null;
  other_doc_value?: string | null;
  is_match: boolean;
  discrepancy_note?: string | null;
}

export interface ReviewAICitationItem {
  citation_id: string;
  source_type: string;
  title: string;
  page?: number | null;
  snippet?: string | null;
}

export interface ReviewAIExplanationSection {
  recommendation?: string | null;
  confidence_label?: string | null;
  summary?: string | null;
  strengths: string[];
  concerns: string[];
  review_items: string[];
  grounded_citations: ReviewAICitationItem[];
  disclaimer: string;
  is_stale: boolean;
  is_available: boolean;
}

export interface ReviewDetailResponse {
  review_id: string;
  organization_id: string;
  tender_id: string;
  tender_number: string;
  tender_title: string;
  bid_id: string;
  bid_number: string;
  bidder_legal_name: string;
  trade_name?: string | null;
  bidder_pan?: string | null;
  bidder_gstin?: string | null;
  review_type: ReviewType;
  issue_type_display?: string | null;
  severity: ReviewSeverity;
  status: ReviewStatus;
  title: string;
  reason: string;
  system_finding: Record<string, any>;
  resolution?: ReviewResolution | null;
  resolution_reason?: string | null;
  effective_compliance_status?: string | null;
  claimed_by_name?: string | null;
  claimed_by_id?: string | null;
  resolved_by_name?: string | null;
  resolved_by_id?: string | null;
  resolved_at?: string | null;
  version: number;
  created_at: string;
  updated_at: string;

  requirement_section?: ReviewRequirementSection | null;
  actual_evidence_section?: ReviewActualEvidenceSection | null;
  source_document_section?: ReviewSourceDocumentSection | null;
  verification_section?: ReviewVerificationEvidenceSection | null;
  compliance_section?: ReviewComplianceEvidenceSection | null;
  risk_section?: ReviewRiskSection | null;
  clarification_section?: ReviewClarificationSection | null;
  cross_document_section: CrossDocumentComparisonRow[];
  ai_explanation_section?: ReviewAIExplanationSection | null;
  notes_history: ReviewNoteItem[];
}
